### ISSUE TITLE
Implicit dependency works differently when using (`.map` + `.get()`) and `.flatMap` on a `TaskProvider`

### ISSUE THREAD

**Issue by ArchangelX360** [gradle/gradle#25645]
Consider the following `build.gradle.kts`:
```kotlin
abstract class MyTask @Inject constructor(
    objectFactory: ObjectFactory
): DefaultTask() {
    @get:Internal
    val someDirectory = objectFactory.directoryProperty()

    @get:OutputFile
    val myFile = someDirectory.map { d -> d.file("file") }

    @TaskAction
    fun execute() {
        myFile.get().asFile.writeText("coucou")
    }
}

val t1Provider = tasks.register<MyTask>("t1") {
    someDirectory.set(layout.buildDirectory)
}

val someDir = layout.buildDirectory.dir("someDir")

tasks.register<Sync>("t2") {
    from(t1Provider.flatMap { it.myFile })
    into(someDir)
}

tasks.register<Sync>("t3") {
    from(t1Provider.map { it.myFile.get() })
    into(someDir)
}
```

### Expected Behavior

- `./gradlew t2` runs `t1` and `t2`
- `./gradlew t3` runs `t1` and `t3`

OR

- `./gradlew t2` runs `t2`
- `./gradlew t3` runs `t3`

but consistency at least.

Note that the [docs of `.flatMap`](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Provider.html#flatMap-org.gradle.api.Transformer-) mentions:
> this method offers a convenient way of connecting together task inputs and outputs
> [...]
> An added benefit of connecting input and output properties like this is that Gradle can automatically detect task dependencies based on such connections.

so probably the first expected behaviour would be the most expected by users.

### Current Behavior

- `./gradlew t2` runs `t2`
- `./gradlew t3` runs `t1` and `t3`

### Steps to Reproduce

- add the snippet provided in `Expected Behavior` section
- Run `./gradlew t2`
- Run `./gradlew t3`

### Gradle version

8.2

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team and is prioritized by them.

---

The problem here is that `myFile` output property is. mapped from an internal one and Gradle misses the dependency in that case. That is what makes the `flatMap`'s case wrong.

There are at least two ways to mitigate it: 
1. Fix the behavior to capture a reference to the task for `flatMap` when an output property is derived from an internal property. 
2. Task validation can throw an error saying it is not supported.

**Comment by big-guy** [comment]
I think there's a simpler way to define the task:
```
abstract class MyTask : DefaultTask() {
    @get:OutputFile
    abstract val myFile: RegularFileProperty

    @TaskAction
    fun execute() {
        myFile.get().asFile.writeText("coucou")
    }
}
```
Gradle will provide the appropriate implementation for `getMyFile`. This also separates the conventions of the task (where myFile is written relative to some other directory) from the work of the task. 

And then later, you can specify exactly where you want the output to go:
```
val t1Provider = tasks.register<MyTask>("t1") {
    myFile.set(layout.buildDirectory.file("file"))
}
```
And then for `t2` or `t3`, you have a couple of options:
```
val someDir = layout.buildDirectory.dir("someDir")

tasks.register<Sync>("t2") {
    from(t1Provider.flatMap { it.myFile })
    into(someDir)
}

tasks.register<Sync>("t3") {
    from(t1Provider)
    into(someDir)
}
```
Here `t3` is copying all of the outputs from `t1` and `t2` is only copying the explicit output `myFile`. This only makes a difference if `t1` produced two different outputs and you wanted to distinguish them.

**Comment by ArchangelX360** [comment]
@big-guy The code you propose is not a viable workaround as it changes the public API of `MyTask` compared to the original example.

> This also separates the conventions of the task (where myFile is written relative to some other directory) from the work of the task.

Both the change of API and the content of the previous quote is precisely what we *don't* want in my use-case.
The definition of `myFile` as `Provider` is done on purpose, as in my use-case the task's caller must not be allowed to set `myFile`.
`myFile` location is only known by `MyTask` as it relates to the internals of the task's execution.
Changing the location would break the build thus `myFile` should not be exposed as a `Property`.

**Comment by abdulowork** [comment]
Hi. I want to second that this bug is very much not obvious and it makes me second guess myself every time I want to rely on implicit task dependencies with `flatMap`. I often want to use `flatMap`, but I instead end up doing the:
```kotlin
task.map { it.property.get() }
```
so that my tasks dependencies don't spontaneously vanish from the `ConfigurableFileCollection` that I am building using those providers.

In the discussion it was also noted that:
> output property is mapped from an internal one and Gradle misses the dependency in that case

I want to point out that the behavior seems to also depends on the type signature of `OutputFile` property. For example the following carries the task dependency just fine:
```kotlin
abstract class Producer : DefaultTask() {
    @get:Internal
    val internal = project.objects.fileProperty()
    @get:OutputFile
    val output: Property<RegularFile> = internal
}

val aProducer = tasks.register("aP", Producer::class.java).flatMap { it.output }
val aConsumer = tasks.register("aC", Consumer::class.java) { consumer.set(aProducer) }
```

but changing `val output: Property<RegularFile>` to `val output: Provider<RegularFile>` is enough to break the task dependency from `aC` to `aP`. This is also reproducible without the `Internal` property; for example the following doesn't carry the task dependency if flatMapped to `output`:
```kotlin
abstract class BrokenProducer : DefaultTask() {
    @get:OutputFile
    val output: Provider<RegularFile> = project.objects.fileProperty()
    // e.g. you might have output file instantiated within the task
    // val output: Provider<RegularFile> = project.layout.buildDirectory.file("my_output")
}
```

**Comment by Vampire** [comment]
It probably indeed would be nice if such providers would preserver task dependencies.

But just as a work-around, while this still changes the public API of the task technically, it does not semantically and the task dependency does work if you declare the property like this:
```kotlin
val myFile = objectFactory
    .fileProperty()
    .value(someDirectory.map { d -> d.file("file") })
    .apply { finalizeValue() }
```

**Comment by rkrisztian** [comment]
Here's a full example in Groovy DSL for the workaround that @Vampire suggested (https://github.com/gradle/gradle/issues/25645#issuecomment-2634613927), if anyone finds it useful (tested with Gradle 8.13):

```groovy
abstract class Producer extends DefaultTask {
    // THIS LINE DOES NOT EXECUTE THE PRODUCER
//    @OutputFile final Provider<RegularFile> outputFile = project.layout.buildDirectory.file('file.txt')

    // THIS IS THE WORKAROUND
    @OutputFile final RegularFileProperty outputFile = project.objects.fileProperty()
            .value(project.layout.buildDirectory.file('file.txt'))
            .tap { finalizeValue() }

    // OR THIS IF YOU PREFER:
//    @OutputFile abstract RegularFileProperty getOutputFile()
//
//    Producer() {
//        outputFile.value(project.layout.buildDirectory.file('file.txt'))
//                .tap { finalizeValue() }
//    }

    @TaskAction void produce() {
        outputFile.get().asFile.createNewFile()
    }
}

abstract class Consumer extends DefaultTask {
    @InputFile abstract RegularFileProperty getInputFile()

    @TaskAction void consume() {
        logger.quiet "Input file: '${inputFile.get()}'"
    }
}

def producer = tasks.register('producer', Producer)

tasks.register('consumer', Consumer) {
    inputFile = producer.flatMap { it.outputFile }
}
```

**Comment by Vampire** [comment]
Yeah, the important part is, that it has to be a `Property` formally.
If you would change `RegularFileProperty` to `Provider<RegularFile>` and keep the rest like it is, it would also not work iirc as only `Property` output fields get the task dependencies propagated.
